src/nupic/algorithms/TemporalMemory.cpp

```diff
@@ -0,0 +1,1093 @@
+  }
+
+  Int i = _rng.getUInt32((UInt32)leastUsedCells.size());
+  return leastUsedCells[i];
+}
+
+vector<Synapse> TemporalMemory::activeSynapsesForSegment(
+  Segment& segment,
+  set<Cell>& activeCells,
+  Connections& _connections)
+{
+  vector<Synapse> synapses;
+
+  for (Synapse synapse : _connections.synapsesForSegment(segment))
+  {
+    SynapseData synapseData = _connections.dataForSynapse(synapse);
+
+    if (find(activeCells.begin(), activeCells.end(),
+      synapseData.presynapticCell) != activeCells.end())
+    {
+      synapses.push_back(synapse);
+    }
+  }
+  return synapses;
+}
+
+void TemporalMemory::adaptSegment(
+  Segment& segment,
+  vector<Synapse>& activeSynapses,
+  Connections& _connections)
+{
+  for (Synapse synapse : _connections.synapsesForSegment(segment))
+  {
+    SynapseData synapseData = _connections.dataForSynapse(synapse);
+    Permanence permanence = synapseData.permanence;
+
+    if (find(activeSynapses.begin(), activeSynapses.end(),
+      synapse) != activeSynapses.end())
+      permanence += permanenceIncrement_;
+    else
+      permanence -= permanenceDecrement_;
+
+    // Keep permanence within min / max bounds
+    //permanence = max(0.0, min(1.0, permanence));
+    if (permanence > 1.0)
+      permanence = 1.0;
+    if (permanence < 0.0)
+      permanence = 0.0;
+
+    _connections.updateSynapsePermanence(synapse, permanence);
+  }
+}
+
+set<Cell> TemporalMemory::pickCellsToLearnOn(
+  Int iN,
+  Segment& segment,
+  set<Cell>& _winnerCells,
+  Connections& _connections)
+{
+  vector<Cell> candidates(_winnerCells.begin(), _winnerCells.end());
+
+  // Remove cells that are already synapsed on by this segment
+  for (auto synapse : _connections.synapsesForSegment(segment))
+  {
+    SynapseData synapseData = _connections.dataForSynapse(synapse);
+    Cell presynapticCell = synapseData.presynapticCell;
+
+    if (find(candidates.begin(), candidates.end(),
+      presynapticCell) != candidates.end())
+    {
+      candidates.erase(find(candidates.begin(), candidates.end(), presynapticCell));
+    }
+  }
+
+  // Pick n cells randomly
+  Int n = min(iN, (Int)candidates.size());
+  sort(candidates.begin(), candidates.end());
+
+  set<Cell> cells;
+  for (int c = 0; c < n; c++)
+  {
+    Int i = _rng.getUInt32((UInt32)candidates.size());
+    cells.insert(candidates[i]);
+    candidates.erase(find(candidates.begin(), candidates.end(), candidates[i]));
+  }
+
+  return cells;
+}
+
+Int TemporalMemory::columnForCell(Cell& cell)
+{
+  _validateCell(cell);
+
+  return cell.idx / cellsPerColumn_;
+}
+
+vector<Cell> TemporalMemory::cellsForColumn(Int column)
+{
+  _validateColumn(column);
+
+  Int start = cellsPerColumn_ * column;
+  Int end = start + cellsPerColumn_;
+
+  vector<Cell> cellsInColumn;
+  for (Int i = start; i < end; i++)
+  {
+    cellsInColumn.push_back(Cell(i));
+  }
+
+  return cellsInColumn;
+}
+
+Int TemporalMemory::numberOfColumns(void)
+{
+  Int acc(1);
```

Comment:
Yes, but I tried to stick to the Py as close as possible to easy review. I've changed this to return that member variable, and inlined the function proto.